### PR TITLE
FEATURE: Allow List for PMs

### PR DESCRIPTION
**Pull request by oblakeerickson**
This feature adds a new user setting that is disabled by default that
allows them to specify a list of users that are allowed to send them
private messages. This way they don't have to maintain a large list of
users they don't want to here from and instead just list the people they
know they do want. Staff will still always be able to send messages to
the user.

### CODE REVIEW

File: app/services/user_updater.rb [hunk 1]
```diff
@@ -225,6 +230,28 @@ def update_ignored_users(usernames)
...
+    usernames ||= ""
+    desired_usernames = usernames.split(",").reject { |username| user.username == username }
+    desired_ids = User.where(username: desired_usernames).where(admin: false, moderator: false).pluck(:id)
+    if desired_ids.empty?
+      AllowedPmUser.where(user_id: user.id).destroy_all
+    else
+      AllowedPmUser.where('user_id = ? AND allowed_pm_user_id not in (?)', user.id, desired_ids).destroy_all
+
+      # SQL is easier here than figuring out how to do the same in AR
+      DB.exec(<<~SQL, now: Time.now, user_id: user.id, desired_ids: desired_ids)
```
Comment (eviltrout) [inline app/services/user_updater.rb:245]:
```suggestion
      DB.exec(<<~SQL, now: Time.zone.now, user_id: user.id, desired_ids: desired_ids)
```

File: lib/post_creator.rb [hunk 2]
```diff
@@ -128,6 +128,23 @@ def valid?
...
       end
 
+      # Is Allowed PM users list enabled for any recipients?
+      users_with_allowed_pms = allowed_pms_enabled(users).pluck(:id).uniq
+
+      # If any of the users has allowed_pm_users enabled check to see if the creator
+      # is in their list
+      if users_with_allowed_pms.any?
+        users_sender_can_pm = allowed_pms_enabled(users)
+          .where("allowed_pm_users.allowed_pm_user_id = #{@user.id.to_i}")
```
Comment (eviltrout) [inline lib/post_creator.rb:138]:
```suggestion
          .where("allowed_pm_users.allowed_pm_user_id" => @user.id.to_i)
```
I *think* the above syntax works and means you needn't bother with string interpolation which is riskier.

File: spec/models/allowed_pm_user_spec.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
```
Comment (eviltrout) [inline spec/models/allowed_pm_user_spec.rb:1]: This file can be removed.

**Comment by oblakeerickson** [comment]
Thanks!

> already present in the `PostCreator` which skips username checks.

Yep, that's correct.

**Pull request merged**